### DISCUSSION TITLE
questions about e_id and node_id on LinkNeighborLoader

### DISCUSSION THREAD

**Discussion by cooperationz** [pyg-team/pytorch_geometric#7233] [Q&A]
1、When I use LinkNeighborLoader for sampling, what are the n_id and input_id of the results returned in each batch?
I think input_id is the index of test_data["user", "rates", "movie"].edge_label, e_id is the index of test_data["user", "rates", "movie"].edge_label, and then these two sets of indexes Get the n_id of user and movie in this batch, but I can't get the expected result from the above code, is my code wrong? 

```
from torch_geometric.data import HeteroData
import torch_geometric.transforms as T
import torch

test_data = HeteroData()
# Save node indices:
test_data["user"].node_id = torch.tensor([0, 1, 3, 4])
test_data["movie"].node_id = torch.tensor([0, 1, 3, 4])
#edge_index
test_data['user', 'rates', 'movie'].edge_index = torch.tensor([[0, 1, 1, 1, 2, 2, 2, 3, 3, 3, 4, 4],
                                  [1, 0, 2, 3, 1, 3, 4, 1, 2, 4, 2, 3]])
#edge_label
test_data['user', 'rates', 'movie'].edge_label = torch.tensor([1,0,1,0])
#edge_label_index
test_data['user', 'rates', 'movie'].edge_label_index = torch.tensor([[0,1,4,3],
                                    [4,2,0,1]])
test_data = T.ToUndirected()(test_data)


from torch_geometric.loader import LinkNeighborLoader
# Define seed edges:
edge_label_index = test_data["user", "rates", "movie"].edge_label_index
edge_label = test_data["user", "rates", "movie"].edge_label

test_data_loader = LinkNeighborLoader(
    data=test_data,
    num_neighbors=[2], 
    edge_label_index=(("user", "rates", "movie"), edge_label_index),
    edge_label=edge_label,
    batch_size=2,

)
# Inspect a sample:
sampled_data = next(iter(test_data_loader))


print(sampled_data['user'],'\n')
print(sampled_data['movie'],'\n')

print(sampled_data['user', 'rates', 'movie'].edge_index,'\n')
print(sampled_data['user', 'rates', 'movie'].edge_label,'\n')
print(sampled_data['user', 'rates', 'movie'].edge_label_index,'\n')
print(sampled_data['user', 'rates', 'movie'].e_id,'\n')
print(sampled_data['user', 'rates', 'movie'].input_id,'\n')

print(test_data["user", "rates", "movie"].edge_label_index[:,sampled_data['user', 'rates', 'movie'].input_id],'\n')
print(test_data["user", "rates", "movie"].edge_index[:,sampled_data['user', 'rates', 'movie'].e_id],'\n')


```

<img width="635" alt="question" src="https://user-images.githubusercontent.com/115331582/234442376-dbecda7f-f361-4266-8241-c7b1ac6d2ea8.png">

2、Also, is node_id the result after mapping? If yes, why not a continuation starting from 0?

1、When I use LinkNeighborLoader for sampling, what are the n_id and input_id of the results returned in each batch?
I think input_id is the index of test_data["user", "rates", "movie"].edge_label, e_id is the index of test_data["user", "rates", "movie"].edge_label, and then these two sets of indexes Get the n_id of user and movie in this batch, but I can't get the expected result from the above code, is my code wrong? 

[https://colab.research.google.com/drive/1xpzn1Nvai1ygd_P5Yambc_oe4VBPK_ZT?usp=sharing](url)

**Comment by rusty1s** [comment]
1. Your understanding is fully correct. Sorry for not making this clearer in the documentation. The mismatch between `node_id` and `n_id` is due to the fact that you initialize it as
```python
test_data["user"].node_id = torch.tensor([0, 1, 3, 4])
```
rather than
```python
test_data["user"].node_id = torch.tensor([0, 1, 2, 3])
```
2. `n_id` refers to the global mapping, and is useful, for example, if you want to query embeddings via `torch.nn.Embedding` for all sampled nodes for example.

**Reply by cooperationz** [reply]
thanks for your reply.
Does the initialization of node_id have to be continuous starting from 0? Will initialization like my code work properly?

**Reply by rusty1s** [reply]
`node_id` can be anything you want it to be, this is up to you. I was just referring to the mismatch in outputs between `node_id` (set by you) and `n_id` (set by us via `torch.arange(data.num_nodes)`).
